keps/sig-cluster-lifecycle/addons/20191013-install-addons-via-kubeadm.md

```diff
@@ -0,0 +1,450 @@
+
+Maintaining these manifests and this logic with kubeadm has to this point has allowed users to recieve a useful, default setup.
+As the ecosystem grows, this is proving to now be too inflexible for users and vendors.
+
+Use of ClusterAPI means less operators will be manually mutating clusters before they are expected to function.
+Kubeadm does not need to maintain CNI installation, but users should be able to use kubeadm to produce a functioning cluster, complete with with operational CNI. 
+
+### Goals
+
+<!-- List the specific goals of the KEP.
+How will we know that this has succeeded? -->
+
+Integrating the addon-installer library and API with kubeadm will allow users to opt out of installing kube-proxy or CoreDNS/kube-dns.
+Some reasons for this may be:
+- The user does not desire pod networking and/or service-discovery in their cluster.
+- CNI implementations such as kube-router and Cilium can actually replace kube-proxy.
+
+The "core" addons can be referenced and installed from a central, extensible, shared implementation.
+In the case of CoreDNS, this may even make sense to be managed by the upstream project instead of maintaining copies of manifests in kubernetes/kubernetes.
+
+This integration will also allow users to extend these "core" addons using their own kustomize driven patches.
+Patches are expected to be supplied via the API, filesystem, or git with additional potential for OCI images.
+
+This will also enable users to specify additional addons to install such as CNI implementations, drivers, webhook-servers, or other API's and operators.
+
+Simple upgrades to newer manifests for a given addon should be supported.
+Removal should also be supported.
+
+The mechanisms and API's for doing this are independently factored from kubeadm meaning other installers such as kops and eksctl as well as cloud and vendor solutions may implement their own defaulting logic and user-interfaces.
+
+### Non-Goals
+
+Addons are currently expected to be applied in order with no success condition beyond API validity.
+No dependency system or DAG is currently proposed or in-scope.
+
+Discussions of how addons may validate ComponentConfigs and other files and how the AddonInstaller API can enable that are welcome. There are existing efforts in this area already (currently championed by @rosti) that may need to be cross-referenced.
+
+Usable operators for addons such as the "CoreDNS Operator" are for consideration regarding the motivation, UX, and vision of this KEP's features. However, implementation, packaging, and delivery of these operators is out of scope for the technical implementation of the AddonInstaller in kubeadm.
+
+Providing an "uninstall" signal, hook, or mechanism for addons is currently out of scope but is open for discussion.
+
+## Proposal
+
+An alpha feature gate "`AddonInstaller`" will be added to kubeadm; it will default to false may be user-enabled using the exisiting flag and `ClusterConfiguration`.
```

Comment:
```suggestion
An alpha feature gate "`AddonInstaller`" will be added to kubeadm; it will default to false may be user-enabled using the existing flag and `ClusterConfiguration`.
```